dashboard/app/tree.go

```diff
@@ -0,0 +1,725 @@
+		// Demand that at least one of the finished jobs has finished successfully.
+		return pollResultSkip{}
+	}
+	// Since we have a repro for it, it definitely crashed at some point.
+	perNode[ctx.repoNode] = pollResultDone{Crashed: true}
+	allLabels := append(ctx.selectRepoLabels(true, perNode), ctx.selectRepoLabels(false, perNode)...)
+	for _, label := range allLabels {
+		if label == ctx.repoNode.repo.LabelIntroduced || label == ctx.repoNode.repo.LabelReached {
+			// It looks like our reproducer does not work on other trees.
+			// Just in case verify that it still works on the original one.
+			result := ctx.runRepro(ctx.repoNode.repo, wantNewAny(lastDone), runOnHEAD{})
+			resultDone, ok := result.(pollResultDone)
+			if !ok {
+				return result
+			}
+			if !resultDone.Crashed {
+				// Unfortunately the repro no longer works. Don't assign labels.
+				return pollResultSkip{}
+			}
+		}
+	}
+	var labels []BugLabel
+	for _, label := range allLabels {
+		labels = append(labels, BugLabel{Label: OriginLabel, Value: label})
+	}
+	ctx.bug.SetLabels(makeLabelSet(ctx.c, ctx.bug.Namespace), labels)
+	return pollResultSkip{}
+}
+
+// selectRepoNodes attributes bugs to trees depending on the patch testing results.
+func (ctx *bugTreeContext) selectRepoLabels(in bool, results map[*repoNode]pollTreeJobResult) []string {
+	crashed := map[*repoNode]bool{}
+	for node, result := range results {
+		done, ok := result.(pollResultDone)
+		if ok {
+			crashed[node] = done.Crashed
+		}
+	}
+	for node := range crashed {
+		if !crashed[node] {
+			continue
+		}
+		// (1) The in = true case:
+		// If, for a tree X, there's a tree Y from which commits flow to X and the reproducer crashed
+		// on Y, X cannot be among bug origin trees.
+		// (1) The in = false case:
+		// If, for a tree X, there's a tree Y to which commits flow to X and the reproducer crashed
+		// on Y, X cannot be the last tree to which the bug has spread.
+		for otherNode := range node.reachable(!in) {
+			crashed[otherNode] = false
+		}
+	}
+	ret := []string{}
+	for node, set := range crashed {
+		if !set {
+			continue
+		}
+		if in && node.repo.LabelIntroduced != "" {
+			ret = append(ret, node.repo.LabelIntroduced)
+		} else if !in && node.repo.LabelReached != "" {
+			ret = append(ret, node.repo.LabelReached)
+		}
+	}
+	return ret
+}
+
+// Test if there's any sense in testing other trees.
+// For example, if we hit a bug on a mainline, there's no sense to test linux-next to check
+// if it's a linux-next bug.
+func (ctx *bugTreeContext) labelsCanBeSet() bool {
+	for node := range ctx.repoNode.reachable(true) {
+		if node.repo.LabelIntroduced != "" {
+			return true
+		}
+	}
+	for node := range ctx.repoNode.reachable(false) {
+		if node.repo.LabelReached != "" {
+			return true
+		}
+	}
+	return ctx.repoNode.repo.LabelIntroduced != "" ||
+		ctx.repoNode.repo.LabelReached != ""
+}
+
+func (ctx *bugTreeContext) missingBackports() pollTreeJobResult {
+	if !ctx.repoNode.repo.MissingBackports {
```

Comment:
Shouldn't we also check bug.HasUserLabel here?

Reply:
Indeed. I've added a check.